libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/ConwayFeatures.hs

```diff
@@ -368,36 +378,58 @@ testGov pf = do
               )
             ]
         )
-    eitherEpochState1 = runEPOCH (EPOCH pf) epochState0 (EpochNo 2) poolDistr
-    epochState1 = expectRight "Error running runEPOCH: " eitherEpochState1
-    ledgerState2 = epochState1 ^. esLStateL
-    constitution = (ensConstitution . rsEnactState) (epochState1 ^. esLStateL . lsUTxOStateL . utxosGovStateL . cgRatifyStateL)
-  assertEqual "constitution after enactment" constitution (proposedConstitution @era)
-
+    -- Wait two epochs
+    !eitherEpochState1 = runEPOCH (EPOCH pf) epochState0 (EpochNo 2) poolDistr
```

Comment:
Why all the bangs?

Reply:
It makes it easier to debug when the test fails. This makes the let statements evaluate in sequence so I can just add `traceM` between them.

Reply:
> The only thing bang patterns here do is say: regardless if the binding is used or not, evaluate it.

TIL :raised_hands: :bow: :100:

Reply:
I agree that it's a bad assumption in the sense that it doesn't specify *when* the value has to be evaluated, but it seems to be a good enough assumption for testing and tracing, because GHC seems to be evaluating them as if they were monadic binds. See this: https://play.haskell.org/saved/EzsvMLPx

Reply:
@Soupstraw You are totally wrong about this!!

Without running the code, try to guess which error message you will see first:

```haskell
main :: IO ()
main = do
  let !x = error "x"
      !z = x + y
      !y = error "y"
  traceM "Still made it"
```

https://play.haskell.org/saved/YbjO28OE

So, please remove those bangs, so not to confuse anyone.

Reply:
If you ever want to require specific ordering for debugging or testing, you should either:
* manually specify it with a `seq`, eg. 
```haskell
x `seq` y `seq` traceM "Tested"
```
* or use [`evaluate`](https://downloads.haskell.org/ghc/9.2.8/docs/html/libraries/base-4.16.4.0/Control-Exception.html#v:evaluate):
```haskell
x' <- evaluate x
y' <- evaluate y
traceM "Tested"
```

Reply:
Ok you convinced me :smile:

Reply:
but maybe it's a good idea to use `evaluate` on `runEPOCH` and `runLEDGER` since those can fail

Reply:
What I suggest is to change `expectRight` into:

```haskell
expectRight :: (HasCallStack, Show a) => String -> Either a b ->  IO b
expectRight msg = either (\x -> error (msg <> show x)) evaluate
```